src/internals/strings/impl/length.ts

```diff
@@ -1,66 +1,44 @@
 // Original impl https://gist.github.com/sno2/7dac868ec6d11abb75250ce5e2b36041
 
 import { Add } from "../../numbers/impl/addition";
-import { Mul } from "../../numbers/impl/multiply";
+import { StringIterator as StrIter } from "./utils";
 
-type TCache = [string, 0[]];
+// Implementation of the following algorithm:
+//
+// function length(s, acc) {
+//   if (s === "") {
+//       return 0;
+//   } else if (s.indexOf(acc) === 0) {
+//       return acc.length + length(s.slice(acc.length), acc + acc);
+//   } else {
+//       return length(s, acc.slice(acc.length / 2));
+//   }
+// }
 
 type LengthUp<
-  T extends string,
-  $Acc extends 0[] = [],
-  $Cache extends TCache[] = [[`$${string}`, [0]]]
-> =
-  //
-  $Cache extends [infer C extends TCache, ...infer $RestCache extends TCache[]]
-    ? (
-        `${C[0]}${C[0]}_` extends `$${string}$${infer $After}`
-          ? `${C[0]}${$After}` extends `${infer $Before}_`
-            ? $Before
-            : never
-          : never
-      ) extends infer $DoubleC extends string
-      ? `$${T}` extends `${$DoubleC}${infer $Rest}`
-        ? $Cache["length"] extends 12 // 2^13 is the last block size within the complexity limit
-          ? LengthDownSlow<
-              $Rest,
-              Add<$Acc["length"], Mul<C[1]["length"], 2>>,
-              [[$DoubleC, [...C[1], ...C[1]]], ...$Cache]
-            >
-          : LengthUp<
-              $Rest,
-              [...$Acc, ...C[1], ...C[1]],
-              [[$DoubleC, [...C[1], ...C[1]]], ...$Cache]
-            >
-        : `$${T}` extends `${C[0]}${infer $Rest}`
-        ? LengthUp<$Rest, [...$Acc, ...C[1]], $Cache>
-        : LengthDown<T, $Acc, $RestCache>
-      : never
-    : $Acc["length"];
+  Str extends string,
+  $Length extends number | bigint = 0,
+  It extends StrIter.Iterator = StrIter.Init
+> = It extends StrIter.Iterator
+  ? StrIter.Double<It> extends infer $DoubleIt extends StrIter.Iterator
+    ? StrIter.CutAt<Str, $DoubleIt> extends `${infer $Rest}`
+      ? StrIter.Size<It> extends 12 // 2^13 is the last block size within the complexity limit
```

Comment:
Sorry, did not notice this before but you actually don't need this limitation anymore because the block size was referring to the `0[]` which I stored the counts in. However, you have switched it to only using the string number math so you can just do the else case here.

Reply:
Actually without it, we can't handle length of more than ~100k while with it we can go to ~1M.
So i'm ok with this trick